java/src/main/java/ai/rapids/cudf/PinnedMemoryPool.java

```diff
@@ -0,0 +1,292 @@
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package ai.rapids.cudf;
+
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.PriorityQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/**
+ * This provides a pool of pinned memory similar to what RMM does for device memory.
+ */
+public final class PinnedMemoryPool implements AutoCloseable {
+  private static final Logger log = LoggerFactory.getLogger(PinnedMemoryPool.class);
+  private static final long ALIGNMENT = 8;
+
+  // These static fields should only ever be accessed when class-synchronized.
+  // Do NOT use singleton_ directly!  Use the getSingleton accessor instead.
+  private static volatile PinnedMemoryPool singleton_ = null;
+  private static ExecutorService initService = null;
+  private static Future<PinnedMemoryPool> initFuture = null;
+
+  private long pinnedPoolBase;
+  private PriorityQueue<MemorySection> freeHeap = new PriorityQueue<>(new SortedBySize());
+  private int numAllocatedSections = 0;
+  private long availableBytes;
+
+  private static class SortedBySize implements Comparator<MemorySection> {
+    @Override
+    public int compare(MemorySection s0, MemorySection s1) {
+      // We want the largest ones first...
+      int ret = Long.compare(s1.size, s0.size);
+      if (ret == 0) {
+        ret = Long.compare(s0.baseAddress, s1.baseAddress);
+      }
+      return ret;
+    }
+  }
+
+  private static class MemorySection {
+    private long baseAddress;
+    private long size;
+
+    MemorySection(long baseAddress, long size) {
+      this.baseAddress = baseAddress;
+      this.size = size;
+    }
+
+    boolean canCombine(MemorySection other) {
+      boolean ret = (other.baseAddress + other.size) == baseAddress ||
+          (baseAddress + size) == other.baseAddress;
+      log.trace("CAN {} COMBINE WITH {} ? {}", this, other, ret);
+      return ret;
+    }
+
+    void combineWith(MemorySection other) {
+      assert canCombine(other);
+      log.trace("COMBINING {} AND {}", this, other);
+      this.baseAddress = Math.min(baseAddress, other.baseAddress);
+      this.size = other.size + this.size;
+      log.trace("COMBINED TO {}\n", this);
+    }
+
+    MemorySection splitOff(long newSize) {
+      assert this.size >= newSize;
```

Comment:
do we have a need for a splitOff memory section to have the same size as the original? Should we instead only check if ```this.size > newSize```